engines/lab/engine.cpp

```diff
@@ -0,0 +1,1190 @@
+ * This code is based on Labyrinth of Time code with assistance of
+ *
+ * Copyright (c) 1993 Terra Nova Development
+ * Copyright (c) 2004 The Wyrmkeep Entertainment Co.
+ *
+ */
+
+#include "lab/lab.h"
+
+#include "lab/anim.h"
+#include "lab/dispman.h"
+#include "lab/eventman.h"
+#include "lab/image.h"
+#include "lab/intro.h"
+#include "lab/labsets.h"
+#include "lab/music.h"
+#include "lab/processroom.h"
+#include "lab/resource.h"
+#include "lab/tilepuzzle.h"
+#include "lab/utils.h"
+
+namespace Lab {
+
+// LAB: Labyrinth specific code for the special puzzles
+#define SPECIALLOCK         100
+#define SPECIALBRICK        101
+#define SPECIALBRICKNOMOUSE 102
+
+enum Items {
+	kItemHelmet = 1,
+	kItemBelt = 3,
+	kItemPithHelmet = 7,
+	kItemJournal = 9,
+	kItemNotes = 12,
+	kItemWestPaper = 18,
+	kItemWhiskey = 25,
+	kItemLamp = 27,
+	kItemMap = 28,
+	kItemQuarter = 30
+};
+
+#define kCondLampOn         151
+#define kCondBeltGlowing     70
+#define kCondUsedHelmet     184
+
+enum Monitors {
+	kMonitorMuseum = 71,
+	kMonitorGramophone = 72,
+	kMonitorUnicycle = 73,
+	kMonitorStatue = 74,
+	kMonitorTalisman = 75,
+	kMonitorLute = 76,
+	kMonitorClock = 77,
+	kMonitorWindow = 78,
+	//kMonitorBelt = 79,
+	kMonitorLibrary = 80,
+	kMonitorTerminal = 81
+	//kMonitorLevers = 82
+};
+
+enum MainButtons {
+	kButtonPickup,
+	kButtonUse,
+	kButtonOpen,
+	kButtonClose,
+	kButtonLook,
+	kButtonInventory,
+	kButtonLeft,
+	kButtonForward,
+	kButtonRight,
+	kButtonMap
+};
+
+enum AltButtons {
+	kButtonMainDisplay,
+	kButtonSaveLoad,
+	kButtonUseItem,
+	kButtonLookAtItem,
+	kButtonPrevItem,
+	kButtonNextItem,
+	kButtonBreadCrumbs,
+	kButtonFollowCrumbs
+};
+
+static char initcolors[] = { '\x00', '\x00', '\x00', '\x30',
+							 '\x30', '\x30', '\x10', '\x10',
+							 '\x10', '\x14', '\x14', '\x14',
+							 '\x20', '\x20', '\x20', '\x24',
+							 '\x24', '\x24', '\x2c', '\x2c',
+							 '\x2c', '\x08', '\x08', '\x08' };
+
+uint16 LabEngine::getQuarters() {
+	return _inventory[kItemQuarter]._quantity;
+}
+
+void LabEngine::setQuarters(uint16 quarters) {
+	_inventory[kItemQuarter]._quantity = quarters;
+}
+
+void LabEngine::drawRoomMessage(uint16 curInv, CloseDataPtr closePtr) {
+	if (_lastTooLong) {
+		_lastTooLong = false;
+		return;
+	}
+
+	if (_alternate) {
+		if ((curInv <= _numInv) && _conditions->in(curInv) && !_inventory[curInv]._bitmapName.empty()) {
+			if ((curInv == kItemLamp) && _conditions->in(kCondLampOn))
+				// LAB: Labyrinth specific
+				drawStaticMessage(kTextkLampOn);
+			else if (_inventory[curInv]._quantity > 1) {
+				Common::String roomMessage = _inventory[curInv]._name + "  (" + Common::String::format("%d", _inventory[curInv]._quantity) + ")";
+				_graphics->drawMessage(roomMessage.c_str(), false);
+			} else
+				_graphics->drawMessage(_inventory[curInv]._name.c_str(), false);
+		}
+	} else
+		drawDirection(closePtr);
+
+	_lastTooLong = _graphics->_lastMessageLong;
+}
+
+void LabEngine::freeScreens() {
+	for (uint16 i = 0; i < 20; i++) {
```

Comment:
In general I would suggest to use "int" for everything which does not need to have its range bounded specifically.